src/coreclr/jit/compiler.cpp

```diff
@@ -2305,50 +2309,66 @@ void Compiler::compSetProcessor()
+        // there exists no hardware which doesn't support the entire feature set. To simplify
+        // the overall JIT implementation, we currently require the entire set of ISAs to be
+        // supported and disable AVX512 support otherwise.
+
+        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512F));
+        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512F_VL));
         assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512BW));
+        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512BW_VL));
         assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512CD));
+        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512CD_VL));
         assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512DQ));
-        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512F));
-        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512F_VL));
+        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512DQ_VL));
 
         instructionSetFlags.AddInstructionSet(InstructionSet_Vector512);
-    }
-    else
-    {
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512F);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512F_VL);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512BW);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512BW_VL);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512CD);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512CD_VL);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512DQ);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512DQ_VL);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512VBMI);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512VBMI_VL);
 
-#ifdef TARGET_AMD64
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512F_X64);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512F_VL_X64);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512BW_X64);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512BW_VL_X64);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512CD_X64);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512CD_VL_X64);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512DQ_X64);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512DQ_VL_X64);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512VBMI_X64);
-        instructionSetFlags.RemoveInstructionSet(InstructionSet_AVX512VBMI_VL_X64);
-#endif // TARGET_AMD64
+        if (preferredVectorByteLength == 0)
+        {
+            CORINFO_XARCH_CPU xarchCpuInfo;
```

Comment:
I am not sure how I feel about passing the full CPU details to the JIT to compute this bit. It is too revealing abstraction.

Would it be better for this bit to be computed by the EE and passed to the JIT as a `CORJIT_FLAG_AVX512_THROTTLING` bit? We have older flags like that for x86 (`CORJIT_FLAG_USE_CMOV`, `CORJIT_FLAG_USE_FCOMI`, etc.).

Reply:
> It is too revealing abstraction.

Not sure I understand the concern here, it's something that most native compilers carry around as information.

> Would it be better for this bit to be computed by the EE and passed to the JIT as a CORJIT_FLAG_AVX512_THROTTLING bit?  We have older flags like that for x86 (CORJIT_FLAG_USE_CMOV, CORJIT_FLAG_USE_FCOMI, etc.).

Those are more like a legacy form of the newer `InstructionSet` flags than what we have today. They're 32-bit only and given our other requirements (such as SSE/SSE2) could/should be removed entirely at this point, they don't serve any real purpose and are always going to be set.

This is really just exposing some core information about the machine to the compiler which allows it to make educated decisions on how to generate code. Most of the time its not a factor we want to consider, but it makes it easy (and won't require a new flag or JIT/EE change each time) if a relevant scenario does come up.

This also avoids any messiness around having `PreferredVectorWidth` on the JIT side but the flag which determines whether to throttle or not on the VM side.

Reply:
How should this work with the earlier ask?
> This config should be in the JIT so that it can be controlled by the --codegenopt switch for AOT compilers, similar to how other undocumented optimization settings can be controlled.

Are we keeping `PreferredVectorWidth` as a JIT flag and using it in combination with the VM passing down any `CORJIT_FLAG_AVX512_THROTTLING`?

That is, on the JIT side are we changing this to:
```
if (PreferredVectorWidth == 0)
{
    if (HasInstructionSet(Avx512F)
    {
        if (CORJIT_FLAG_AVX512_THROTTLING)
        {
            PreferredVectorWidth = 256;
        }
        else
        {
            PreferredVectorWidth = 512;
        }
    }
    else if (HasInstructionSet(Avx2))
    {
        PreferredVectorWidth = 256;
    }
    else
    {
        PreferredVectorWidth = 128;
    }
}
```

The native VM will then populate `CORJIT_FLAG_AVX512_THROTTLING` based on the `CPUID info` while crossgen/naot will default to having it clear (with the assumption more people will be on newer hardware over time, and power users can use the undocumented switch)?

Reply:
> The native VM will then populate CORJIT_FLAG_AVX512_THROTTLING based on the CPUID info while crossgen/naot will default to having it clear (with the assumption more people will be on newer hardware over time, and power users can use the undocumented switch)?

Yep.

Reply:
Fixed.